File: bencher/tests/c/qsort/qsort.c
#include "qsort.h"

long long partition(SortData_t data, long long low, long long high) {
    SortItem_t temp;
    SortItem_t pivot = data[high];
    long long i = (low - 1);

    for (SortItem_t j = low; j <= high - 1; j++) {
        if (data[j] <= pivot) {
            i++;
            temp = data[i];
            data[i] = data[j];
            data[j] = temp;
        } 
    }

    temp = data[i + 1];
    data[i + 1] = data[high];
    data[high] = temp;

    return (i + 1); 
}

void quick_sort(SortData_t data, long long low, long long high) {
    if (low < high) {
        long long pi = partition(data, low, high);

        quick_sort(data, low, pi - 1);
        quick_sort(data, pi + 1, high);
    }
}

void data_print(SortData_t data, unsigned int size) {
    for(int i = 0; i < size; i++) {
        printf("%llu\n", data[i]);
    }
}

File: bencher/tests/c#/qsort/qsort.cs
namespace qsort;

internal abstract class QuickSort
{
    private static long Partition(ulong[] data, long low, long high)
    {
        var pivot = data[high];
        var i = low - 1;

        for (var j = low; j <= high - 1; j++)
        {
            if (data[j] <= pivot)
            {
                i++;
                (data[i], data[j]) = (data[j], data[i]);
            }
        }

        (data[i + 1], data[high]) = (data[high], data[i + 1]);

        return i + 1;
    }

    public static void Sort(ulong[] data, long low, long high)
    {
        if (low < high) {
            var pi = Partition(data, low, high);

            Sort(data, low, pi - 1);
            Sort(data, pi + 1, high);
        }
    }

    public static void DataPrint(ulong[] data, uint size)
    {
        for (var i = 0; i < size; i++)
        {
            Console.Write($"{data[i]} ");
        }
    }
}